pkg/controller/clone-controller.go

```diff
@@ -0,0 +1,264 @@
+	pvcInformer, podInformer cache.SharedIndexInformer
+	cloneImage               string
+	pullPolicy               string // Options: IfNotPresent, Always, or Never
+	verbose                  string // verbose levels: 1, 2, ...
+}
+
+func NewCloneController(client kubernetes.Interface, pvcInformer, podInformer cache.SharedIndexInformer, cloneImage string, pullPolicy string, verbose string) *CloneController {
+	c := &CloneController{
+		clientset:     client,
+		pvcQueue:      workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		podQueue:      workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		pvcInformer:   pvcInformer,
+		podInformer:   podInformer,
+		cloneImage:    cloneImage,
+		pullPolicy:    pullPolicy,
+		verbose:       verbose,
+	}
+
+	// Bind the pvc SharedIndexInformer to the pvc queue
+	c.pvcInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			key, err := cache.MetaNamespaceKeyFunc(obj)
+			if err == nil {
+				c.pvcQueue.AddRateLimited(key)
+			}
+		},
+		// this is triggered by an update or it will also be
+		// be triggered periodically even if no changes were made.
+		UpdateFunc: func(old, new interface{}) {
+			key, err := cache.MetaNamespaceKeyFunc(new)
+			if err == nil {
+				c.pvcQueue.AddRateLimited(key)
+			}
+		},
+	})
+
+	// Bind the pod SharedIndexInformer to the pod queue
+	c.podInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			key, err := cache.MetaNamespaceKeyFunc(obj)
+			if err == nil {
+				c.podQueue.AddRateLimited(key)
+			}
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			
+			//TEMPORARY - for debug -  to be delete
+			//tmp, _ := json.Marshal(&newObj))
+		    //glog.V(Vdebug).Info(fmt.Println(string(tmp)))
+			///////////
+			
+			key, err := cache.MetaNamespaceKeyFunc(newObj)
+			if err == nil {
+				c.podQueue.AddRateLimited(key)
+			}
+		},
+	})
+
+	return c
+}
+
+func (c *CloneController) Run(threadiness int, stopCh <-chan struct{}) error {
+	defer func() {
+		c.pvcQueue.ShutDown()
+		c.podQueue.ShutDown()
+	}()
+	glog.V(Vadmin).Infoln("Starting clone controller Run loop")
+	if threadiness < 1 {
+		return errors.Errorf("expected >0 threads, got %d", threadiness)
+	}
+
+	if !cache.WaitForCacheSync(stopCh, c.pvcInformer.HasSynced) {
+		return errors.New("Timeout waiting for pvc cache sync")
+	}
+	if !cache.WaitForCacheSync(stopCh, c.podInformer.HasSynced) {
+		return errors.New("Timeout waiting for pod cache sync")
+	}
+	glog.V(Vdebug).Infoln("CloneController cache has synced")
+
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(c.runPVCWorkers, time.Second, stopCh)
+		go wait.Until(c.runPodWorkers, time.Second, stopCh)
+	}
+	<-stopCh
+	return nil
+}
+
+func (c *CloneController) runPodWorkers() {
+	for c.ProcessNextPodItem() {
+	}
+}
+
+func (c *CloneController) runPVCWorkers() {
+	for c.ProcessNextPvcItem() {
+	}
+}
```

Comment:
From approx L35 to here: is it possible to create a generic cdi controller "factory" func that takes care of all of this boilerplate code so that we don't repeat the same stuff 3 times? @davidvossel @copejon ?

Reply:
@jeffvance It would be an improvement, but is outside the scope of this PR.  A large scale refactor of how we generate these controllers will have to take place.

@zvikorn I see your point, it's maybe a shame that go doesn't support inheritance, even in a simplistic way.  Implementing a factory may fit our needs here, but again, not in this PR.

Reply:
Go implements the ‘has-a’ relationship that we can do here and will solve everything.
But maybe I don’t see the whole picture ;-)

Reply:
I think we will want to revisit this later.  For now, let's work to get the feature merged and we can do a better refactor using go interfaces later.